Fix and test macOS Debugger

thestr4ng3r opened the pull request:
# DO NOT SQUASH

 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

First simple tests to prevent at least complete breakage.

**Test plan**

CI should pass

ret2libc commented on .github/workflows/ci.yml:
```diff
@@ -210,12 +210,10 @@ jobs:
+      - name: Disable user authentication for debugging (macOS)
         if: matrix.os == 'macos-11' && matrix.enabled
         run: |
-          security authorizationdb read system.privilege.taskport > taskport.plist
```
Isn't this necessary anymore?

thestr4ng3r replied:
It is, but it can be simplified to not having to edit the plist manually. The `sudo security authorizationdb write system.privilege.taskport allow` opens it up further but it doesn't matter since we have root anyway.

XVilka reviewed (changes requested):
```

[XX] db/archos/darwin-x64/dbg dcu + read
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'dcu main
pi 8 @ rip
' bins/mach0/hello-objc-osx
-- stdout
--- expected
+++ actual
@@ -1,4 +1,3 @@
-push rbp
 mov rbp, rsp
 sub rsp, 0x10
 mov rax, qword [section.13.__DATA.__objc_classrefs]
@@ -6,3 +5,4 @@
 lea rax, reloc.objc_msgSend_fixup
 mov rax, qword [rax]
 lea rsi, reloc.objc_msgSend_fixup
+call rax
-- stderr
Continue until 0x100000bf8
EXC_SOFTWARE
```

thestr4ng3r committed:
Remove XNU_USE_PTRACE

The ptrace-based backend is broken and has been disabled by default
since 2016.

thestr4ng3r committed:
De-include xnu_threads.c and xnu_excthreads.c

With XNU_USE_PTRACE removed, these source files can now be compiled as
standalone objects unconditionally.

thestr4ng3r committed:
Add very basic debug tests for macOS

The system.privilege.taskport change in CI could also be simplified without
having to manually edit a plist.

thestr4ng3r committed:
Fix attaching in XNU Debugger

ptrace with PT_ATTACHEXC will send a SIGSTOP to the process, which we then
sometimes received in the first mach_msg loop while waiting for hitting a
breakpoint, when having been already attached at this point.
See also bsd/kern/mach_process.c:287 in xnu-8020.121.3.

Correct attaching works like this, in exactly this order:
* posix_spawn the process with POSIX_SPAWN_START_SUSPENDED
* set up the Mach port for receiving exceptions
* ptrace(PT_ATTACHEXC) -> sends SIGSTOP
* wait on the Mach port until receiving the SIGSTOP msg
* call task_suspend
* reply to the msg

Then we have a beautiful process suspended right at the start in dyld.

thestr4ng3r committed:
Improve logging for EXC_SOFTWARE on XNU, showing Unix signals

thestr4ng3r committed:
Fix detaching in XNU Debugger

PT_DETACH must always be called, even when killing the process.

thestr4ng3r committed:
Remove unused variable in xnu_debug.c

XVilka commented on librz/debug/p/native/xnu/xnu_excthreads.c:
```diff
@@ -374,14 +400,20 @@ static int handle_exception_message(RzDebug *dbg, exc_msg *msg, int *ret_code) {
-static int __xnu_wait(RzDebug *dbg, int pid) {
-	// here comes the important thing
+/**
+ * Wait for a mach exception, reply to it and handle it.
```
Shouldn't it be "Mach"?

thestr4ng3r replied:
Yes

XVilka reviewed (approved):
`XNU_USE_PTRACE` was broken even more, right?

thestr4ng3r commented:
> XNU_USE_PTRACE was broken even more, right?

I tried it and nothing was working with it. From what I know the approach is only useful for something like Mac OS X <= 10.2, which is extremely old, when the exceptions were not yet sent through Mach IPC. But right now rizin does not even compile there with debugger enabled, so should be done from scratch later if really needed.

thestr4ng3r committed:
Fix rz_xnu_get_cur_thread() messing up the target's mappings

Getting the thread list with task_threads() allocates a mapping in the
caller's memory that should then be freed with vm_deallocate() later.
However rz_xnu_get_cur_thread() was passing the debuggee's task to the
deallocate rather than its own, meaning the thread list would probably
leak and the debuggee's maps would get broken if there was something at
that address.
